File: 9.go
package main

import (
	"./itertools"
	"fmt"
	"io/ioutil"
	"math"
	"os"
	"strconv"
	"strings"
)

func sumInPreamble(preamble []int, sum int) bool {
	c := make(chan []int)
	go itertools.Combinations(preamble, 2, c)
	for x := range c {
		if x[0]+x[1] == sum {
			return true
		}
	}
	return false
}

func main() {
	bytes, err := ioutil.ReadFile("9.in")
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	lines := strings.Split(strings.TrimSpace(string(bytes)), "\n")

	input := make([]int, len(lines))
	for i, x := range lines {
		xi, _ := strconv.Atoi(x)
		input[i] = xi
	}

	// Part 1
	var faulty int
	for i := 25; i < len(input); i++ {
		preamble := input[i-25 : i]
		if !sumInPreamble(preamble, input[i]) {
			faulty = input[i]
		}
	}
	fmt.Println(faulty)

	// Part 2
	var min, max, sum int
	for i := 0; i < len(input); i++ {
		sum = 0
		max = 0
		min = math.MaxInt32
		for _, x := range input[i:] {
			if x < min {
				min = x
			}
			if x > max {
				max = x
			}
			sum += x
			if sum >= faulty {
				break
			}
		}
		if sum == faulty {
			break
		}
	}
	fmt.Println(min + max)
}

File: 9.py
from itertools import combinations


def sum_in_preamble(pre, x):
    for a, b in combinations(pre, 2):
        if a + b == x:
            return True

    return False


if __name__ == "__main__":
    with open("9.in") as f:
        lines = [x.strip() for x in f.readlines()]

    input = [int(x) for x in lines]

    for i in range(25, len(input)):
        preamble = input[i - 25 : i]
        if not sum_in_preamble(preamble, input[i]):
            faulty = input[i]
            break

    print(faulty)

    for i in range(len(input)):
        numbers = []
        for x in input[i:]:
            numbers.append(x)
            if sum(numbers) >= faulty:
                break

        if sum(numbers) == faulty:
            break

    print(min(numbers) + max(numbers))
